### ISSUE TITLE
Expert parallelism and "Unimplemented group_offset support" for jax.lax.ragged_dot

### ISSUE THREAD

**Issue by pcmoritz** [jax-ml/jax#34168]
Dear all,

Recently I have been trying to implement expert parallelism with Jax (primarily on GPUs but ideally the same code would also work on TPUs). As a small twist, LoRA support is pretty important for me since I'm working on a Tinker compatible OSS backend, see https://github.com/NovaSky-AI/SkyRL/tree/main/skyrl-tx

I had two possible strategies in mind:

- Either use `jax.lax.ragged_all_to_all` to distribute the hidden states to different ranks according to which expert they should be computed on and then call `jax.lax.ragged_dot` on the local experts on each rank
- Or broadcast all hidden states to all the ranks and then use `jax.lax.ragged_dot` with a `group_offset` of the group that corresponds to the first local expert

My plan was to implement both of these solutions and evaluate them on (a) performance and (b) simplicity. However, while trying to implement the second solution (see https://github.com/NovaSky-AI/SkyRL/pull/842), I was running into the error `Unimplemented group_offset support` for `jax.lax.ragged_dot` -- it looks like that parameter is not implemented on GPUs (but probably is on TPUs but I currently have no access to those so I can't test it).

I wonder if this functionality could be implemented for GPUs (that's why I made this issue a **feature request**) or if there are any general recommendations or advice on how to best implement expert parallelism with Jax (there is surprisingly little public information available). Maybe there is other strategies I haven't thought about, or the above strategies can be implemented in a particularly simple way.

There is https://github.com/sgl-project/sglang-jax/pull/439 which seems to already have profiled both solution and found the second one to be faster (I assume that was on TPUs) and I also found https://github.com/jax-ml/jax-llm-examples/blob/e236f634b455805a524769a4b3098f4a76a2a8d3/llama4/llama4_jax/model.py#L1006 which is a little more complex.

All the best and thanks for your help,
Philipp.

**Comment by agolajko** [comment]
@jakevdp would love to hear if there are any recommendations on the best approach here.

Any guidance on whether GPU support for `jax.lax.ragged_dot` with `group_offset` is on the roadmap, or if there's a preferred pattern for this kind of workload, would be really helpful.
Thanks!

**Comment by pcmoritz** [comment]
I was digging into this some more. One option to make `group_offset` happen would be to modify https://github.com/rdyro/gpu_ragged_dot/blob/main/gpu_ragged_dot.py

I think for vanilla jax, the codepath that is taken is https://github.com/jax-ml/jax/blob/fd57ef614d56a846a50cd2d897fd6aa9a4d096b3/jax/_src/lax/lax.py#L6350 and from there probably into some XLA codepath that might not be super easy to modify without more XLA knowledge. But it would be very beneficial to do it, because (a) then we wouldn't need custom autotuning to get good performance and (b) it would make programs more portable between GPUs and TPUs :)

**Comment by voznesenskym** [comment]
Hey! @piotrfilipiuk is a good authority on ragged dot. 

That said:

> I was running into the error Unimplemented group_offset support for jax.lax.ragged_dot -- it looks like that parameter is not implemented on GPUs

This is just a not implemented at the jax layer, we don't have support for it atm, but it shouldn't be too rough to add.

The ragged_dot impl on TPU (I am most familiar with this one, my apologies for not answering the GPU side, as I wrote the jvp and transpose for this op, and the XLA lowering for the bwd) is implemented via a pattern match and lowering to a mosaic kernel written in mosaic IR. I am not sure what the gpu path is -  but for the tpu path:

```
return c->AddInstruction(HloInstruction::CreateConcatenate(
    concat_shape,
    {c->AddInstruction(HloInstruction::CreateConstant(
         LiteralUtil::CreateR1<int32_t>({0}))), 
     group_ends},
```
That 0 is the hardcoded groupoffset (very readable code, I know lol) so we would have to upgrade the CreateGroupOffsets and  to take a base offset, and then plumb it via BuildMetadataModule to the MemRefStoreOp we use to go from constant 0 to use that same base_offset - and a loop w/ an addop over group_sizes should handle it - the matmul wont need to change, I think we just need to shift the data in the group_offsets.

For the GPU path, someone should correct me here, but I believe there is not actually a proper ragged dot, but rather we fall through to a padded dot?

**Comment by pcmoritz** [comment]
Hey @voznesenskym thanks a lot for answering, that is great to know (and thanks a ton for walking us throught the TPU code path!). I didn't realize group_offset was not implemented on TPUs, but now that you point it out it makes a lot of sense. I assumed it was working because sglang-jax used it, but now I realize that it actually calls into a custom kernel (https://github.com/primatrix/sglang-jax/blob/main/python/sgl_jax/srt/kernels/gmm/megablox_gmm_kernel/gmm.py#L286) instead of using ragged_dot.

The reason why this would be great to have is it allows to implement sharded experts and yet keep static shapes for jax (but not pay for the extra computation). My plan is to get a simple implementation working that just uses jax.lax.ragged_dot without group_offset under the hood but exposes the interface with group_offset (and just sends the extra tokens to experts that are available in the shard). Happy to share that once it is finished (I did some experiments with this and it is actually already faster than just doing TP sharding on the experts). I'll also investigate which kernels are good to do it and I can also report back on that if I find anything useful out.

Would love to learn more about which code path(s) are taken for the GPU implementation of jax.lax.ragged_dot if anybody has that information.

Also if there is more guidance on the best ways to do expert parallelism / sharding with jax I'd love to know!

**Comment by bchetioui** [comment]
> For the GPU path, someone should correct me here, but I believe there is not actually a proper ragged dot, but rather we fall through to a padded dot?

That's correct. There's a [`RaggedDotRewriter`](https://github.com/openxla/xla/blob/9ae3d6dab2c10c8195c8d9862f475904c7cdca91/xla/hlo/transforms/expanders/ragged_dot_rewriter.cc#L347-L369) pass that expands ragged dot into a `dot_general`. There is an experimental flag `xla_gpu_experimental_use_ragged_dot_fusion` to bypass this expander and to emit code for a proper ragged dot, but I don't believe the underlying support was ever landed (cc @vwbaker to contradict me if that is wrong).

**Comment by pcmoritz** [comment]
Thanks a lot for all your feedback! I have implemented the workaround of using `jax.lax.ragged_dot` and masking out the extra tokens now in https://github.com/NovaSky-AI/SkyRL/pull/860/files (in case this is useful for others who want to do similar things too) and implemented the expert sharding with it using `shard_map` in https://github.com/NovaSky-AI/SkyRL/pull/842.

It performs surprisingly well actually, and reduces the step time from about 110s (if I just use TP sharding) down to 40s (using EP sharding) in my concrete case, this is on 8xH100. However for larger batch sizes I expect implementing `group_offset` natively will give a large speedup since it avoids a lot of extra work. I wonder if you have any recommendations on what the best kernels (especially for GPUs) are.

So far I have found:

- https://github.com/sgl-project/sglang-jax/blob/main/python/sgl_jax/srt/kernels/gmm/megablox_gmm_kernel/gmm.py
- the kernels in https://github.com/AI-Hypercomputer/maxtext, 
- https://github.com/rdyro/gpu_ragged_dot/blob/main/gpu_ragged_dot.py
- the pallas kernels (https://github.com/jax-ml/jax/blob/main/jax/experimental/pallas/ops/gpu/ragged_dot_mgpu.py, https://github.com/jax-ml/jax/blob/main/jax/experimental/pallas/ops/gpu/blackwell_ragged_dot_mgpu.py)

There is also https://github.com/openxla/tokamax, but it seems like these also don't support group_offset.

**Comment by voznesenskym** [comment]
If I am reading it right, right now you're doing:

[all m tokens] → ragged_dot (computes ALL tokens) → mask away the ones you don't care about
Which is fine! but you're paying for compute you throw away.

Have you tried:

[slice out only the tokens you care about] → ragged_dot (smaller input) ?

**Comment by pcmoritz** [comment]
Thanks for your answer @voznesenskym ! I'd love to do exactly that, but the shape of the tokens I care about will be different in each invocation (and each layer) and it is dynamic based on the expert routing (and in the extreme case, all tokens could go to one expert, though in practice that's very unlikely). My understanding is that such dynamic shapes are not supported, at least with JIT which I need to get sufficient performance. I have seen some people define a "capacity" (e.g. 1.2 times of the number of tokens you would get if everything is balanced) and then throwing away the extra tokens, but that doesn't seem like a good strategy to me, it is important to keep correctness.

If there is a way to do the slicing and expert routing dynamically efficiently, I'd love to know :)

**Comment by voznesenskym** [comment]
Yeah, good point! I think at that point you'd have to reach for pallas
